patches/tModLoader/Terraria.ModLoader.Recipes/RecipeBuilder.cs

```diff
@@ -0,0 +1,184 @@
+			for (int i = 0; i < types.Length; i++)
+				Requires(types[i]);
+
+
+			return this;
+		}
+
+
+		/// <summary>Adds an ingredient to this recipe with the given item name from the given mod, and with the given stack stack. If the mod parameter is null, then it will automatically use an item from the mod creating this recipe.</summary>
+		/// <param name="mod">The mod.</param>
+		/// <param name="itemName">Name of the item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">The item " + itemName + " does not exist in mod " + mod.Name + ". If you are trying to use a vanilla item, try removing the first argument.</exception>
+		public RecipeBuilder Requires(Mod mod, string itemName, int stack = 1)
+		{
+			_recipe.AddIngredient(mod ?? _recipe.mod, itemName, stack);
+
+			return this;
+		}
+
+
+		/// <summary>Adds an ingredient to this recipe of the given type of item and stack size.</summary>
+		/// <param name="modItem">The item.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		public RecipeBuilder Requires(ModItem modItem, int stack = 1)
+		{
+			_recipe.AddIngredient(modItem, stack);
+
+			return this;
+		}
+
+
+		/// <summary>Adds a recipe group ingredient to this recipe with the given RecipeGroup name and stack size. Vanilla recipe groups consist of "Wood", "IronBar", "PresurePlate", "Sand", and "Fragment".</summary>
+		/// <param name="recipeGroup">The name.</param>
+		/// <param name="stack">The stack.</param>
+		/// <returns></returns>
+		/// <exception cref="RecipeException"></exception>
+		public RecipeBuilder Requires(string recipeGroup, int stack = 1)
+		{
+			_recipe.AddRecipeGroup(recipeGroup, stack);
+
+			return this;
+		}
+
+
+		/// <summary>
+		/// Adds one or many required crafting station(s) with the given tile type(s) to the recipe being built.
+		/// Ex.:
+		/// <example>At(TileID.WorkBenches, TileID.Anvils)</example>
+		/// </summary>
+		/// <param name="tileTypes"></param>
+		/// <returns></returns>
+		public RecipeBuilder At(params int[] tileTypes)
+		{
+			for (int i = 0; i < tileTypes.Length; i++)
+				_recipe.AddTile(tileTypes[i]);
+
+			return this;
+		}
+
+
+		public RecipeBuilder RequiresLava() => NeedVar(recipe => recipe.needLava = true);
+
+		public RecipeBuilder RequiresHoney() => NeedVar(recipe => recipe.needHoney = true);
+
+		public RecipeBuilder RequiresWater() => NeedVar(recipe => recipe.needWater = true);
+
+		public RecipeBuilder RequiresSnowBiome() => NeedVar(recipe => recipe.needSnowBiome = true);
+
+
+		private RecipeBuilder NeedVar(Action<ModRecipe> need)
+		{
+			need(_recipe);
+
+			return this;
+		}
+
+
+		/// <summary>Adds this recipe to the game. Call this after you have finished setting the result, ingredients, etc.</summary>
+		/// <returns></returns>
+		/// <exception cref="RecipeException">A recipe without any result has been added.</exception>
+		public void Build() => _recipe.AddRecipe();
```

Comment:
You should have a 'finalized' or 'added' bool on this builder and check against it in any of the method calls, the builder is 'done' once Build() is called.

Reply:
Thats a lot of hand-holding. Do we need to throw an error every time as well ?